File: field-force-java/src/Connection/Packet.java
package Connection;

import AI.AI;
import Board.Consumable;
import Board.Fire;
import Board.Wall;
import Game.*;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

import static java.lang.System.exit;

public class Packet {

    private List<Byte> buffer;
    private byte[] readableBuffer;
    private int readPos;

    public Packet() {
        buffer = new ArrayList<>();
        readPos = 0;
    }

    public Packet(byte[] data) {
        buffer = new ArrayList<>();
        readPos = 0;

        setBytes(data);
    }

    private void setBytes(byte[] data) {
        write(data);
        readableBuffer = toByteArray();
    }

    /**
     * @return
     * @throws Exception
     * reads 4 bytes and increased read pos
     */
    public int readInt() throws Exception {
        if (buffer.size() > readPos)
        {
            int value = convertToInt(splitMessage(readableBuffer, readPos, readPos+3));
            readPos += 4;
            return value;
        }
        else
        {
            throw new Exception("Could not read value of type 'int'!");
        }
    }

    /**
     * @return
     * @throws Exception
     * reads 4 bytes for the length of the string
     * reads the amount of bytes that are read in the length
     */
    public String readString() throws Exception {
        try
        {
            int length = readInt();
            String message = new String(toByteArray(), readPos, length);
            if (message.length() > 0)
            {
                readPos += length;
            }
            return message;
        }
        catch (Exception e)
        {
            throw new Exception("Could not read value of type 'string'!");
        }
    }

    /**
     * @param dimension
     * @return
     * @throws Exception
     * reads all the chars for the map and puts in appropriate 2d array
     */
    public char[][] readMap(int dimension) throws Exception {
        char[][] map = new char[dimension][dimension];
        for (int y = 0; y < dimension; y++)
        {
            for (int x = 0; x < dimension; x++)
            {
                map[x][y] = (char) readInt();
            }
        }
        return map;
    }

    /**
     * @return
     * @throws Exception
     * reads playerifnromation and returns a list of player objects
     */
    public List<Player> readPlayers() throws Exception {
        List<Player> players = new ArrayList<>();
        int playerCount = readInt();
        for (int i = 0; i < playerCount; i++){
            int playerNumber = readInt();
            String playerName = readString();
            int hp = readInt();
            int shield = readInt();
            int xPos = readInt();
            int yPos = readInt();
            Skill skill1 = readSkill();
            Skill skill2 = readSkill();
            players.add(new Player((char)(playerNumber+'0'), playerNumber, playerName, hp, shield, xPos, yPos, skill1, skill2));        }
        return players;
    }

    /**
     * @return
     * @throws Exception
     * reads the order in which players are supposed to play
     * index 0 is next player to move
     */
    public List<Integer> readPlayerInTurn() throws Exception {
        List<Integer> playerInTurn = new ArrayList<>();
        int playerCount = readInt();
        for (int i = 0; i < playerCount; i++){
            int playerNumber = readInt();
            playerInTurn.add(playerNumber);
        }
        return playerInTurn;
    }

    /**
     * @return
     * @throws Exception
     * reads all the fires that are available on the map
     */
    public List<Fire> readFires() throws Exception {
        List<Fire> fires = new ArrayList<>();
        int fireCount = readInt();
        for (int i = 0; i < fireCount; i++)
        {
             int x = readInt();
             int y = readInt();
             int duration = readInt();
             fires.add(new Fire('f',x,y, duration));
        }
        return fires;
    }

    /**
     * @return
     * @throws Exception
     * reads all the walls that are available on the map
     */
    public List<Wall> readWalls() throws Exception {
        List<Wall> walls = new ArrayList<>();
        int wallCount = readInt();
        for (int i = 0; i < wallCount; i++)
        {
            int x = readInt();
            int y = readInt();
            int hp = readInt();
            walls.add(new Wall('-', x, y, hp));
        }
        return walls;
    }

    /**
     * @throws Exception
     * reads the full config needed to play the game
     */
    public void readConfig() throws Exception {
        // read basics
        GameConstants.HP = readInt();
        GameConstants.SHIELD = readInt();

        GameConstants.ATTACK_DAMAGE = readInt();
        GameConstants.WALK_IN_PLAYER_DAMAGE = readInt();
        GameConstants.PLAYER_WALKED_INTO_DAMAGE = readInt();

        GameConstants.FIRE_DURATION_ON_MAP = readInt();
        GameConstants.ON_FIRE_EFFECT_DURATION = readInt();
        GameConstants.ON_FIRE_DAMAGE = readInt();

        GameConstants.WALL_HP = readInt();
        GameConstants.WALK_IN_WALL_DAMAGE = readInt();
        GameConstants.WALL_TAKE_DAMAGE = readInt();

        // read consumeables
        Consumable.allConsumables = readConfigConsumables();

        // read skills
        Skill.allSkills = readConfigSkills();
    }

    /**
     * @return
     * @throws Exception
     * reads all valid consumables and puts in global list GameConstants.VALID_CONSUMABLES
     */
    public List<Consumable> readConfigConsumables() throws Exception {
        List<Consumable> consumables = new ArrayList<>();
        int consumableCount = readInt();
        for (int i = 0; i < consumableCount; i++)
        {
            char id = (char) readInt();
            String conName = readString();
            int healing = readInt();
            int shield = readInt();
            consumables.add(new Consumable(id, conName, healing, shield));
            GameConstants.VALID_CONSUMEABLES.add(id);
        }
        return consumables;
    }

    /**
     * @return
     * @throws Exception
     * reads all the consumables that are available on the map
     */
    public List<Consumable> readConsumables() throws Exception {
        List<Consumable> consumables = new ArrayList<>();
        int consumableCount = readInt();
        for (int i = 0; i < consumableCount; i++)
        {
            char id = (char) readInt();
            int x = readInt();
            int y = readInt();
            consumables.add(new Consumable(id, x, y));
        }
        return consumables;
    }

    /**
     * @return
     * @throws Exception
     * reads all valid skills
     */
    public List<Skill> readConfigSkills() throws Exception {
        List<Skill> skills = new ArrayList<>();
        int skillCount = readInt();
        for (int i = 0; i < skillCount; i++)
        {
            int id = readInt();
            String name = readString();
            int cooldown = readInt();
            int range = readInt();
            int value = readInt();
            int type = readInt();
            skills.add(new Skill(id, name, cooldown, range, value, SkillType.fromInt(type)));
        }
        return skills;
    }

    /**
     * @return
     * @throws Exception
     * reads a move object
     */
    public Move readMove() throws Exception {
        MoveType type =  MoveType.fromInt(readInt());
        Direction direction = Direction.fromInt(readInt());
        Skill skill = readSkill();
        return new Move(type, direction, skill);
    }

    /**
     * @return
     * @throws Exception
     * reads a single skill object
     */
    public Skill readSkill() throws Exception {
        int skillId = readInt();
        if(skillId == -1){
            return null;
        }
        int cooldownLeft = readInt();
        return new Skill(skillId, cooldownLeft);
    }

    /**
     * @param data
     * function writes bytes in byte array
     */
    public void write(byte[] data){
        for(Byte b : data) {
            buffer.add(b);
        }
    }

    /**
     * @param value
     * function writes int value as bytes in byte array
     */
    public void write(int value){
        for(Byte b : intToByteArray(value)) {
            buffer.add(b);
        }
    }

    /**
     * @param value
     * function writes float value as bytes in byte array
     */
    public void write(float value){
        for(Byte b : floatToByteArray(value)) {
            buffer.add(b);
        }
    }

    /**
     * @param value
     * function writes string as bytes in byte array
     */
    public void write(String value){
        write(value.length());
        for(Byte b : value.getBytes(StandardCharsets.US_ASCII)) {
            buffer.add(b);
        }
    }

    /**
     * @param move
     * function writes move object as bytes in byte array
     */
    public void write(Move move){
        write(move.getType().getId());
        write(move.getDirection().getId());
        write(move.getSkill());
    }

    /**
     * @param skill
     * function writes skill object as bytes in byte array
     */
    public void write(Skill skill){
        if(skill != null)
            write(skill.getId());
        else
            write(-1);
    }

    /**
     * prepends the length of current data in first 4 indexes of byte array
     */
    public void writeLength(){
        int i = 0;
        for(Byte b : intToByteArray(buffer.size())) {
            buffer.add(i, b);
            i++;
        }
    }

    public List<Byte> getBuffer() {
        return buffer;
    }

    public byte[] toByteArray() {
        final int n = buffer.size();
        byte[] ret = new byte[n];
        for (int i = 0; i < n; i++) {
            ret[i] = buffer.get(i);
        }
        return ret;
    }

    private byte[] intToByteArray(int value){
        ByteBuffer buf = ByteBuffer.allocate(4);
        buf.putInt(value);
        buf.flip();
        byte[] arr = buf.array();
        return reverse(arr, arr.length);
    }

    private byte[] floatToByteArray(float value){
        ByteBuffer buf = ByteBuffer.allocate(4);
        buf.putFloat(value);
        buf.flip();
        byte[] arr = buf.array();
        return reverse(arr, arr.length);
    }

    public static byte[] reverse(byte[] a, int n)
    {
        byte[] b = new byte[n];
        int j = n;
        for (int i = 0; i < n; i++) {
            b[j - 1] = a[i];
            j = j - 1;
        }
        return b;
    }

    public static byte[] splitMessage(byte[] message, int start, int end) {
        int size=end-start+1;
        byte[] splitted=new byte[size];
        int counter=0;
        for(int i=start;i<=end;i++){
            splitted [counter] = message[i];
            counter++;

        }
        return splitted;
    }

    public static int convertToInt(byte[] buffer){
        int result=0;
        if(buffer.length==4) {
            result = buffer[0] << 24 | (buffer[1] & 0xFF) << 16 | (buffer[2] & 0xFF) << 8 | (buffer[3] & 0xFF);
        }
        else if (buffer.length==2){
            result=(buffer[0] & 0xFF) << 8 | (buffer[1] & 0xFF);
        }
        else{
            exit(-1);
        }
        return Integer.reverseBytes(result);
    }
}

File: field-force-cpp/field-force-cpp/connection/packet.cpp
#ifdef _WIN32
#define _CRTDBG_MAP_ALLOC
#include <stdlib.h>
#include <crtdbg.h>
#ifdef _DEBUG
#ifndef DBG_NEW
#define DBG_NEW new ( _NORMAL_BLOCK , __FILE__ , __LINE__ )
#define new DBG_NEW
#endif
#endif  // _DEBUG
#endif
#include "packet.h"
#include "../game/gameconstants.h"
#include "../game/skill.h"
#include "../board/consumable.h"
#include "../board/fire.h"
#include "../board/wall.h"

#include <iostream>
#include <algorithm>

Packet::Packet() { }

Packet::Packet(std::vector<uint8_t> data)
{
	buffer = data;
}

/**
 * \brief
 * reads 4 bytes and increased read pos
 */
int Packet::read_int()
{
	try
	{

		int value;
		std::memcpy(&value, &buffer[read_pos], sizeof(int));
		read_pos += 4;
		return value;
	}
	catch (...)
	{
		std::cerr << "Couldn't read value of type int!" << std::endl;
	}
}

/**
 * \brief reads 4 bytes for the length of the string.
 * reads the amount of bytes that are read in the length
 */
std::string Packet::read_string()
{
	try
	{
		int length = read_int();
		std::string msg;
		msg.resize(length + 1);
		for(int i = 0; i < length; i++)
		{
			msg[i] = buffer[read_pos + i];
		}
		msg[length] = '\0';
		read_pos += length;
		return msg;
	}
	catch (...)
	{
		std::cerr << "Couldn't read string!" << std::endl;
		return "";
	}
}

/**
 * \brief reads the full config needed to play the game
 */
void Packet::read_config()
{
	HP = read_int();
	SHIELD = read_int();

	ATTACK_DAMAGE = read_int();
	WALK_IN_PLAYER_DAMAGE = read_int();
	PLAYER_WALKED_INTO_DAMAGE = read_int();

	FIRE_DURATION_ON_MAP = read_int();
	ON_FIRE_EFFECT_DURATION = read_int();
	ON_FIRE_DAMAGE = read_int();

	WALL_HP = read_int();
	WALK_IN_WALL_DAMAGE = read_int();
	WALL_TAKE_DAMAGE = read_int();

	ALL_CONSUMABLES = read_config_consumables();

	ALL_SKILLS = read_config_skills();
}

/**
 * \brief reads all valid consumables and puts in global list VALID_CONSUMABLES
 */
std::vector<Consumable> Packet::read_config_consumables()
{
	std::vector<Consumable> cons;
	int con_count = read_int();
	for (int i = 0; i < con_count; i++)
	{
		char id = (char)read_int();
		std::string con_name = read_string();
		int healing = read_int();
		int shield = read_int();
		cons.push_back(Consumable(id, con_name, healing, shield));
	}
	return cons;
}

/**
 * \brief reads all valid skills
 */
std::vector<Skill> Packet::read_config_skills()
{
	std::vector<Skill> skills;
	int skills_count = read_int();
	for (int i = 0; i < skills_count; i++)
	{
		int id = read_int();
		std::string skill_name = read_string();
		int cooldown = read_int();
		int range = read_int();
		int value = read_int();
		int type = read_int();
		skills.push_back(Skill(id, skill_name, cooldown, range, value, static_cast<SkillType>(type)));
	}
	return skills;
}

/**
 * \brief reads playerinformation and returns a list of player objects
 */
std::vector<Player> Packet::read_players()
{
	std::vector<Player> players;
	int player_count = read_int();
	for (int i = 0; i < player_count; i++)
	{
		int player_number = read_int();
		std::string player_name = read_string();
		int hp = read_int();
		int shield = read_int();
		int x_pos = read_int();
		int y_pos = read_int();
		Skill skill1 = read_skill();
		Skill skill2 = read_skill();
		players.push_back(Player((char)player_number, player_number, player_name, hp, shield, x_pos, y_pos, skill1, skill2));
	}
	return players;
}

Skill Packet::read_skill()
{
	int skill_id = read_int();
	if(skill_id == -1)
	{
		return Skill();
	}
	int cooldown_left = read_int();
	return Skill(skill_id, cooldown_left);
}

/**
 * \brief reads all the chars for the map and puts in appropriate 2d array
 */
std::vector<std::vector<char>> Packet::read_map(int dimension)
{
	std::vector<std::vector<char>> map;
	map.resize(dimension);
	for (int i = 0; i < dimension; i++)
		map[i].resize(dimension);
	for (int y = 0; y < dimension; y++) 
		for (int x = 0; x < dimension; x++)
			map[x][y] = (char)read_int();
	return map;
}

/**
 * \brief reads the order in which players are supposed to play.
 * index 0 is next player to move
 */
std::vector<int> Packet::read_player_in_turn()
{
	std::vector<int> player_in_turn;
	int player_count = read_int();
	for(int i = 0; i < player_count; i++)
	{
		int player_number = read_int();
		player_in_turn.push_back(player_number);
	}
	return player_in_turn;
}

/**
 * \brief reads all the fires that are available on the map
 */
std::vector<Fire> Packet::read_fires()
{
	std::vector<Fire> fires;
	int fires_count = read_int();
	for (int i = 0; i < fires_count; i++)
	{
		int x = read_int();
		int y = read_int();
		int duration = read_int();
		fires.push_back(Fire('f', x, y, duration));
	}
	return fires;
}

/**
 * \brief reads all the walls that are available on the map
 */
std::vector<Wall> Packet::read_walls()
{
	std::vector<Wall> walls;
	int walls_count = read_int();
	for (int i = 0; i < walls_count; i++)
	{
		int x = read_int();
		int y = read_int();
		int hp = read_int();
		walls.push_back(Wall('-', x, y, hp));
	}
	return walls;
}

/**
 * \brief reads all the consumables that are available on the map
 */
std::vector<Consumable> Packet::read_consumables()
{
	std::vector<Consumable> consumables;
	int cons_count = read_int();
	for (int i = 0; i < cons_count; i++)
	{
		char id = (char)read_int();
		int x = read_int();
		int y = read_int();
		consumables.push_back(Consumable(id, x, y));
	}
	return consumables;
}

/**
 * \brief reads a move object
 */
Move* Packet::read_move()
{
	MoveType type = (MoveType)read_int();
	Direction dir = (Direction)read_int();
	Skill skill = read_skill();
	return new Move(type, dir, skill);
}


/**
 * \brief writes bytes in byte array
 */
void Packet::write(std::vector<uint8_t> value)
{
	for (int i = 0; i < sizeof(value); i++)
	{
		buffer.push_back(value[i]);
	}
}

/**
 * \brief writes int value as bytes in byte array
 */
void Packet::write(int value)
{
	std::vector<uint8_t> data = int_to_byte_array(value);
	for (int i = 0; i < data.size()-1; i++)
	{
		buffer.push_back(data[i]);
	}
}

/**
 * \brief writes string as bytes in byte array
 */
void Packet::write(std::string value)
{
	write((int)value.length());
	for (uint8_t b : value)
	{
		buffer.push_back(b);
	}
}

/**
 * \brief writes move object as bytes in byte array
 */
void Packet::write(Move move)
{
	write(static_cast<int>(move.type));
	write(static_cast<int>(move.direction));
	write(move.skill);
}

/**
 * \brief writes skill object as bytes in byte array
 */
void Packet::write(Skill skill)
{
	write(skill.id);
}

/**
 * \brief prepends the length of current data in first 4 indexes of byte array
 */
void Packet::write_length()
{
	std::vector<uint8_t> length = int_to_byte_array(buffer.size());
	for (int i = 0; i < length.size()-1; i++)
	{
		buffer.insert(buffer.begin() + i, length[i]);
	}
}

std::vector<uint8_t> Packet::int_to_byte_array(int value)
{
	value = reverse_int_byte_array(value);
	std::vector<uint8_t> ret;
	ret.resize(5);
	for (int i = 0; i < 4; i++)
		ret[3 - i] = (value >> (i * 8));
	return ret;
}

std::vector<uint8_t> Packet::to_byte_array()
{
	int n = buffer.size();
	std::vector<uint8_t> ret;
	ret.resize(n);
	for (int i = 0; i < n; i++)
	{
		ret[i] = buffer.at(i);
	}
	return ret;
}

int Packet::reverse_int_byte_array(int data)
{
	return ((data >> 24) & 0xff) |
		((data << 8) & 0xff0000) | 
		((data >> 8) & 0xff00) | 
		((data << 24) & 0xff000000); 
}

int Packet::convert_byte_array_to_int(uint8_t* data)
{
	return int((unsigned char)(data[3]) << 24 |
		(unsigned char)(data[2]) << 16 |
		(unsigned char)(data[1]) << 8 |
		(unsigned char)(data[0]));
}